### PR TITLE
Issue 721 - Content objects for describing request/response payloads

### PR DESCRIPTION
**Pull request by darrelmiller**

### CODE REVIEW

**Comment by zeeshanejaz** [comment]
@darrelmiller is this solely for the purpose of enabling profiles? Sorry I have been late to the party. Can you reference issues that are solved by this. Thanks.

**Comment by ePaul** [comment]
I guess this could also solve at least parts of #146, beside #721?

**Comment by ePaul** [comment]
As already mentioned in [a comment by @fehguy in #721](https://github.com/OAI/OpenAPI-Specification/issues/721#issuecomment-234666646), there is still missing some explanation on how the schema of an `application/x-www-form-urlencoded` (or `multipart/form-data`) representation is mapped to the actual value (or the other way around), i.e. a serialization format.

As this is also missing for other kinds of parameters, I'm okay with addressing it in a separate PR, I just wanted to mention it here again. I guess #665 will have to address that too.

**Comment by darrelmiller** [comment]
@zeeshanejaz The issues addressed by the PR mentioned in the original issue #721. Profiles are simply a way I have found to address the request that some people want to have multiple schemas for the same status code and media type.

**Comment by darrelmiller** [comment]
@ePaul Being able to specify multiple response representations opens the door to supporting #146 but I think it is outside of the spec to try and tie a particular input parameter to the selection of the output representation.  There is still a server implementation behind the API and it is up to the server to decide which representation will be returned.  If the server wants to expose a parameter that directly selects the representation, then it is free to do so.  I don't think the OpenAPI spec needs to get involved at this level.  
As far as generating client code is concerned, the response content type header should provide everything that is needed to deserialize the payload into the appropriate structure.  It is true that most languages don't support the notion of a method that can return multiple different types, so some compromise will need to be made on the signature of calling method.  If having a precise signature is important, then perhaps using multiple paths is a better option.

**Comment by darrelmiller** [comment]
@ePaul I agree that some discussion of how form-data can be described by schemas is important.  But the discussion of how non-JSON data can be described is also bigger than just forms.  Hopefully we can address that subject very soon.

**Comment by fehguy** [comment]
@OAI/tdc let's get your input on this

File: versions/3.0.md [hunk 1]
```diff
@@ -295,16 +292,17 @@ Field Pattern | Type | Description
...
+        "200": {          
           "description": "A list of pets.",
-          "schema": {
-            "type": "array",
-            "items": {
-              "$ref": "#/definitions/pet"
+          "representations" : {
+            "application/json" : {
+                "schema": {
+                            "type": "array",
```
Comment (whitlockjc) [inline versions/3.0.md]: Indentation looks weird when rendered via Markdown.

File: versions/3.0.md [hunk 2]
```diff
@@ -367,24 +365,28 @@ Field Pattern | Type | Description
...
         }
       },
       "default": {
         "description": "error payload",
-        "schema": {
-          "$ref": "#/definitions/ErrorModel"
+        "representations": {
+          "text/html" : {
+            "schema": {
+                      "$ref": "#/definitions/ErrorModel"
```
Comment (whitlockjc) [inline versions/3.0.md]: Indentation again.

**Comment by whitlockjc** [comment]
How do representations support the `void` response, where there is no body?

File: versions/3.0.md [hunk 3]
```diff
@@ -476,42 +477,46 @@ Field Pattern | Type | Description
...
+      "application/x-www-form-urlencoded":{
+        "schema" : {
+           "properties": {
+              "name" 
+                "description" : "Updated name of the pet",
+                "type": "string"
+              "status":
+                "description" : "Updated status of the pet",
+                "type": "string"
+           },
```
Comment (dilipkrish) [inline versions/3.0.md]: It seems like we're going from a flat array of parameters to an object with those fields. Is the intention that we treat the object as an array of form variables?

File: versions/3.0.md [hunk 4]
```diff
@@ -476,42 +477,46 @@ Field Pattern | Type | Description
...
-      "in": "formData",
-      "description": "Updated status of the pet",
-      "required": false,
-      "type": "string"
     }
   ],
+  "requestBody" : {
+    "representations" : {
+      "application/x-www-form-urlencoded":{
+        "schema" : {
```
Comment (dilipkrish) [inline versions/3.0.md]: missing "type" : "object"

File: versions/3.0.md (hunk 3 above)
Comment (darrelmiller) [inline versions/3.0.md]: This is my current proposal for mapping form-style key/value pairs into JSON schema.

**Comment by darrelmiller** [comment]
@whitlockjc I would represent `void` responses by not including a `representations` object.

**Pull request merged**